File: growth/oj/leet_code/algorithms/122-best-time-to-buy-and-sell-stock-ii.py
# -*- coding: utf-8 -*-

'''
TIPS:
    1. operation (buy, sell)
    2. for each (buy, sell), we find the longest incr sequence
    3. but the most important question why longest incr sequence is work?
'''
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if prices is None or len(prices) == 0:
        	return 0

        profit = 0
        longest_incr = None
        buy_in = prices[0]
        i = 1
        while i < len(prices):
        	if prices[i] > prices[i-1]:
        		longest_incr = prices[i]
        		i += 1
        	else:
        		_profit = longest_incr - buy_in if longest_incr else 0
        		if _profit > 0:
        			profit += _profit
        		buy_in = prices[i]
        		longest_incr = None
        		i += 1
        if longest_incr and longest_incr - buy_in > 0:
        	profit += longest_incr - buy_in
        return profit



if __name__ == '__main__':
	obj = Solution()
	print obj.maxProfit(None)
	print obj.maxProfit([])
	print obj.maxProfit([3,3])
	print obj.maxProfit([1,2,3,3,3,4,5])
	print obj.maxProfit([7,1,5,3,6,4])
	print obj.maxProfit([1,2,3,4,5])
	print obj.maxProfit([7,6,4,3,1])
        print obj.maxProfit([1,5,3,6])
        print obj.maxProfit([1,2,3,6])

File: growth/oj/leet_code/algorithms/122-best-time-to-buy-and-sell-stock-ii.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.size() == 0 || prices.size() == 1)
            return 0;
        int total_profit = 0;
        int i = 0;
        while (i < prices.size() -1) {
            while (i < prices.size() - 1 && prices[i] >= prices[i+1])
                i++;
            int v_l = prices[i];

            while ( i < prices.size() - 1 && prices[i] <= prices[i+1])
                i++;
            int v_h = prices[i];

            total_profit += v_h - v_l;
            
        }
        return total_profit;
    }
};


int main() {
    Solution s;
    vector<int> v1{7,1,5,3,6,4}; // -> 7
    cout << s.maxProfit(v1) << endl;

    vector<int> v2{1,2,3,4,5};  // -> 4
    cout << s.maxProfit(v2) << endl;

    vector<int> v3{1,2,3,2,1};  // -> 2
    cout << s.maxProfit(v3) << endl;

    vector<int> v4{1,2,3,2,1,4,5,6,1}; // -> 2 + 5
    cout << s.maxProfit(v4) << endl;


    return 0;
}
